pkg/memory-monitor/src/statistics-collector/main.c

```diff
@@ -0,0 +1,108 @@
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <time.h>
+#include <unistd.h>
+
+#define BUFFER_SIZE 256
+#define OUTPUT_FILE "psi_stats.txt"
+#define DEVICE_OUTPUT_DIR "/persist/memory-monitor/output"
+#define PROC_FILE "/proc/pressure/memory"
+
+void write_to_file(int fd, const char *buffer) {
+    if (write(fd, buffer, strlen(buffer)) == -1) {
+        perror("write");
+        exit(EXIT_FAILURE);
+    }
+    if (fsync(fd) == -1) {
+        perror("fsync");
+        exit(EXIT_FAILURE);
+    }
+}
+
+static void daemonize(void)
+{
+    int pid, sid;
+    pid = fork();
+    if (pid < 0) {
+        exit(EXIT_FAILURE);
+    }
+    // If we got a good PID, then we can exit the parent process
+    if (pid > 0) {
+        exit(EXIT_SUCCESS);
+    }
+
+    // Change the file mode mask
+    umask(0);
+
+    // Create a new SID for the child process
```

Comment:
Why is this necessary?

Reply:
To be detached from the terminal. So we can strat the process in an SSH session and close the terminal, leaving the process collecting the statistics.

Reply:
I thought `setsid` just makes the process to the leader of the process group.
I don't think it installs a `SIGHUP` handler and closes stdin/stdout/stderr.

According to the `man 2 setsid` you still get SIGHUP:
> If  a session has a controlling terminal, and the CLOCAL flag for that terminal is not set, and a terminal hangup > occurs, then the session leader is sent a SIGHUP signal.

Reply:
True, but in the same man 2:
```
Initially, the new session has no controlling terminal.
```